### DISCUSSION TITLE
Can services being deployed using Terraform?

### DISCUSSION THREAD

**Discussion by KaiWalter** [Azure/azure-dev#1595] [Q&A]
So I deploy Azure Container Apps infrastructure with `azd` using Terraform : <https://github.com/KaiWalter/tf-aca>

However I was not able to find documentation how to deploy actual services - like `sender`and `receiver`in my repo - using Terraform and hence reverted to using Bicep. Is this currently the intended way? Deploy infrastructure with Terraform and services still with Bicep?

**Comment by lechnerc77** [comment]
Hi Kai, 

from my understanding it should be possible to create the infrastructure as well as deploy the app by sticking to Terraform without detours to Bicep. 

Technically Azure Container Apps are supported by the Azure Terraform Provider since beginning of February with release 3.43.0. (This might also be the reason why there is not yet an official `azd` template  available for that scenario.) 
Using the resource <https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/container_app> should enable you to define the Azure Container App as you did in your `.bicep` files. The main thing to consider is to add the right labels to your container app to make the connection between receiver and sender defined in the `azure.yaml` file and the resource on Azure when executing the `azd deploy` command. 

Do you stumble across any specific error message during the execution via Terraform?

**Comment by KaiWalter** [comment]
Hey Christian - nice to work on an actual issue together!
I placed a Terraform file which contained a Container App definition (like you suggested) with the exact name of the service - like one would do with Bicep - in the `infra` folder. However on `azp up` or `azd deploy --service ..` CLI was not picking up that file. I wonder what the convention is for the CLI to find the Terraform module describing the service - so what basically is the equivalent to <https://github.com/KaiWalter/tf-aca/tree/main/infra> with Terraform services.

**Reply by lechnerc77** [reply]
The infra folder shouldn't have any influence on the `azd deploy` except that the labels need to be in place. Maybe Azure Container Apps have a bit of a different flow as the code is not directly deployed to the host.
Do you have the original terraform files you mentioned somewhere and can you push them to your repo (maybe in a different branch) - I would like to check them, maybe I find the cause for that behavior.

**Reply by KaiWalter** [reply]
Thanks for also looking into this. With this branch <https://github.com/KaiWalter/tf-aca/tree/tf-apps> I am able to deploy the shell of the Container Apps with a `nginx` image (before the services are build and pushed into ACR).

The first error I got was that `azd up` was looking for a parameters.json file for each service - hence I provided empty ones.

The second error now is that it still tries to find the bicep templates:

```
ERROR: deploying service: deploying service sender package: planning provisioning: planning infrastructure provisioning: creating template: failed to compile bicep template: failed running bicep build: exit code: 1, stdout: , stderr: An error occurred reading file. Could not find file '/home/kai/src/tf-aca/infra/sender.bicep'.
 (exit status 1)
```

**Reply by lechnerc77** [reply]
Just checking your code: the receiver and the sender need to have a tag specified in the `receiver.tf` and `sender.tf` files:

```JSON
"azd-service-nam" : "serviceName"
```

Where service name must be replaced by the one defined in the `azure.yaml` i.e., `sender` and `receiver`.

**Comment by KaiWalter** [comment]
Good suggestion. Right in my Bicep variants always this tag was included. Let's see.

I added the tags to the latest commit, clean up everything and started deploying only infrastructure:

```
azd provision
```

That created the services with the suggested tags
![image](https://user-images.githubusercontent.com/7418323/222154026-5c47d175-5ff7-4b8b-9bee-96a460f077e4.png)

However same behaviour when trying to deploy the service:

```
>azd deploy --service sender

Deploying services (azd deploy)

  (✓) Done: Deploying service sender

ERROR: deploying service: deploying service sender package: planning provisioning: planning infrastructure provisioning: creating template: failed to compile bicep template: failed running bicep build: exit code: 1, stdout: , stderr: An error occurred reading file. Could not find file '/home/kai/src/tf-aca/infra/sender.bicep'.
 (exit status 1)
```

I can see that the container image build succeeded as the environment values show a value for the image name:

```
SERVICE_SENDER_IMAGE_NAME="acrzdi.......m.azurecr.io/tf-aca/sender-kw-tf-aca:azdev-deploy-1............5"
```

**Reply by lechnerc77** [reply]
One step further, so `azd` should know what to deploy where. I think (or hope) that the error message is wrong in the sense that it does not take into account the infrastructure setup type.

This is now a bit "try and error" approach, but  another difference I saw between the Container App (Bicep) sample and your setup is the handling of the image names. The sample:

- defines the image name as output of the corresponding modules
- uses it in the `*.parameters,.json` files

I would therefore recommend to:

-  adjust your setup to transfer the sender and receiver files to two modules that are called form the `main.tf` file. 
- Add the image names as output to the modules (e.g. SERVICE_SENDER_IMAGE_NAME)
- Extract the sender/receiver variable in corresponding files in the modules and default the image name to an empty string, as the nginx is defaulted via the ternary operator in the sender and receiver files
- Add a `main.tfvars.json` with the content

   ```json
   {
       "location": "${AZURE_LOCATION}",
       "environment_name": "${AZURE_ENV_NAME}",
       "principal_id": "${AZURE_PRINCIPAL_ID}"
        "sender_image_name": "${SERVICE_SENDER_IMAGE_NAME}"
       "receiver_image_name":  "${SERVICE_RECEIVER_IMAGE_NAME}"
   }
   ```

**Reply by lechnerc77** [reply]
And another point (that I always forget to do) - add the `--debug` option when executing the `azd deploy`. This should hopefully give more insights where things get stuck.

**Reply by KaiWalter** [reply]
Thanks for the additional inputs @lechnerc77 . However think that all is futile. `azd` is just looking for service Bicep files and not for anything with Terraform flavor. With and without sub-modules, this error is shown.

```
>azd deploy --service sender --debug

  |====   | Deploying service sender2023/03/03 18:07:33 bicep.go:108: using local bicep: /home/kai/.azd/bin/bicep
2023/03/03 18:07:33 service_target_containerapp.go:308: request to show spinner on silent console with message: Creating a deployment plan
2023/03/03 18:07:33 bicep_provider.go:790: Reading parameters template file from: /home/kai/src/tf-aca/infra/sender.parameters.json
2023/03/03 18:07:33 deploy.go:175: - Creating deployment template...
2023/03/03 18:07:33 manager.go:132: Generating Bicep parameters file
2023/03/03 18:07:33 command_runner.go:79: Run exec: '/home/kai/.azd/bin/bicep build /home/kai/src/tf-aca/infra/sender.bicep --stdout'
2023/03/03 18:07:33 manager.go:132: Compiling Bicep template
  |   ====| Deploying service sender2023/03/03 18:07:34 service.go:118: deployed service sender
2023/03/03 18:07:34 deploy.go:175: - Deployment completed...
  (✓) Done: Deploying service sender
```

Let's drop this for the moment.

**Reply by lechnerc77** [reply]
You are right. There is nothing that "convinces" `azd` to use Terraform for the deployment. 
Therefore opened an issue: <https://github.com/Azure/azure-dev/issues/1662>.

**Reply by lechnerc77** [reply]
Update: see <https://github.com/Azure/azure-dev/issues/1662#issuecomment-1515750918>
